### PR TITLE
[ML] Tree to retrain selection

### PR DESCRIPTION
**Pull request by tveasey**
This selects trees with probability proportional to the total loss derivative for training examples for the leaves below each node. This tells you at least locally how much you can improve results by just adjusting the tree predictions (from their current values) and selecting this tree to retrain is analogous to gradient decent. I think however it makes sense to randomise selection since this is a heuristic and it gives us a simple way to refine: randomly sample from this distribution and choose the result which minimises the final training loss. This represents a relatively relatively simple to implement baseline strategy.

This also makes the necessary adjustments to not penalise changing predictions on the new training data. Since this could be out of domain of the original training data this is the right thing to do.

A step towards #1721.

### CODE REVIEW

File: lib/maths/CBoostedTreeUtils.cc [hunk 1]
```diff
@@ -74,9 +101,109 @@ void writeLossCurvature(const TRowRef& row,
...
+
+    auto trainingDataLossGradients = computeLeafLossGradients(trainingDataRowMask);
+
+    for (std::size_t i = 0; i < forest.size(); ++i) {
+        propagateLossGradients(rootIndex(), forest[i], trainingDataLossGradients[i]);
+    }
+    LOG_TRACE(<< "loss gradients = "
+              << core::CContainerPrinter::print(trainingDataLossGradients));
+
+    // We interested in choosing trees for which the total gradient at all
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeUtils.cc:188]:
```suggestion
    // We are interested in choosing trees for which the total gradient at all
```

File: lib/maths/unittest/BoostedTreeTestData.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (valeriy42) [inline lib/maths/unittest/BoostedTreeTestData.h:1]: Good job on refactoring this :+1:

File: lib/maths/unittest/CBoostedTreeUtilsTest.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#include <maths/CBoostedTree.h>
+#include <maths/CBoostedTreeFactory.h>
+#include <maths/CBoostedTreeImpl.h>
+#include <maths/CBoostedTreeLoss.h>
+#include <maths/CBoostedTreeUtils.h>
+
+#include <test/CRandomNumbers.h>
+
+#include "BoostedTreeTestData.h"
+#include "maths/COrderings.h"
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeUtilsTest.cc:19]: This include should be moved up.

File: lib/maths/unittest/CBoostedTreeUtilsTest.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        core::CPackedBitVector allRowsMask{rows, true};
+        std::vector<std::vector<maths::CBoostedTreeNode>> emptyForest;
+        auto probabilities = retrainTreeSelectionProbabilities(
+            1, *frame, impl.extraColumns(), cols - 1, impl.encoder(),
+            allRowsMask, impl.loss(), emptyForest);
+        BOOST_TEST_REQUIRE(probabilities.empty());
+    }
+
+    // Test some invariants of the probabilities.
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeUtilsTest.cc:75]: Can you please elaborate more, which invariants are tested?

**Review by valeriy42 — APPROVED** [review]
Looks good altogether. I have just a few minor comments.

File: lib/maths/CBoostedTreeUtils.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeUtils.cc]: I corrected and also expanded this comment slightly because I think it is a useful reminder. See [ea93475](https://github.com/elastic/ml-cpp/pull/1820/commits/ea934756759014d742ec625bde938a3d8cb24eb0).

File: lib/maths/unittest/CBoostedTreeUtilsTest.cc (hunk 4 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeUtilsTest.cc]: See [ea93475](https://github.com/elastic/ml-cpp/pull/1820/commits/ea934756759014d742ec625bde938a3d8cb24eb0).

File: lib/maths/unittest/CBoostedTreeUtilsTest.cc (hunk 3 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeUtilsTest.cc]: In fact this include was no longer needed so I just removed in [ea93475](https://github.com/elastic/ml-cpp/pull/1820/commits/ea934756759014d742ec625bde938a3d8cb24eb0).
